### PR TITLE
accept :async and :logger options, add a TraceWrapper utility

### PR DESCRIPTION
**Pull request by jfeltesse-mdsol**
See the README about the added/modified options and that utility class.

At the same time I add the standard `bin/` utilities that one would get from bundler upon generating a new gem skeleton.
`bin/console` makes quick prototyping easier.

@jcarres-mdsol @ykitamura-mdsol @ssteeg-mdsol @adriancole

### CODE REVIEW

File: lib/zipkin-tracer/trace_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module ZipkinTracer
+  class TraceWrapper
+    def self.wrap_in_custom_span(config, span_name, span_kind: Trace::Span::Kind::SERVER, app: nil)
```
Comment (ykitamura-mdsol) [inline lib/zipkin-tracer/trace_wrapper.rb]: I think the span kind will be always `SERVER`, maybe it doesn't need to be passed as an argument... 🤔

File: lib/zipkin-tracer/trace_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      raise ArgumentError, "you must provide a block" unless block_given?
+
+      zipkin_config = ZipkinTracer::Config.new(app, config).freeze
+      tracer = ZipkinTracer::TracerFactory.new.tracer(zipkin_config)
+      trace_id = ZipkinTracer::TraceGenerator.new.next_trace_id
+
+      ZipkinTracer::TraceContainer.with_trace_id(trace_id) do
+        tracer.with_new_span(trace_id, span_name) do |span|
```
Comment (ykitamura-mdsol) [inline lib/zipkin-tracer/trace_wrapper.rb]: We need to check `sampled?` here.
Also, some might want to add additional information like in [local-tracing](https://github.com/openzipkin/zipkin-ruby#local-tracing).
How about passing `span` to the `yield`?

```ruby
if trace_id.sampled?
  tracer.with_new_span(trace_id, span_name) do |span|
    span.kind = span_kind
    yield(span)
  end
else
  yield(ZipkinTracer::NullSpan.new)
end
```

File: lib/zipkin-tracer/trace_wrapper.rb (hunk 1 above)
Comment (jfeltesse-mdsol) [reply lib/zipkin-tracer/trace_wrapper.rb]: That's my thinking too but allowing for it to be passed gives the extra flexibility at no cost.
We've already seen people doing wonky things with the library so might be someone will come up with a specific use case where they need to customize the kind.

File: README.md [hunk 3]
```diff
@@ -8,27 +8,34 @@ Rack and Faraday integration middlewares for Zipkin tracing.
...
-* `:annotate_plugin` - plugin function which receives the Rack env, the response status, headers, and body to record annotations
-* `:filter_plugin` - plugin function which receives the Rack env and will skip tracing if it returns false
-* `:whitelist_plugin` - plugin function which receives the Rack env and will force sampling if it returns true
-* `:sampled_as_boolean` - When set to true (default but deprecrated), it uses true/false for the `X-B3-Sampled` header. When set to false uses 1/0 which is preferred.
-* `:trace_id_128bit` - When set to true, high 8-bytes will be prepended to trace_id. The upper 4-bytes are epoch seconds and the lower 4-bytes are random. This makes it convertible to Amazon X-Ray trace ID format v1. (See http://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-request-tracing.html)
+#### Common
+- `:service_name` **REQUIRED** - the name of the service being traced. There are two ways to configure this value. Either write the service name in the config file or set the "DOMAIN" environment variable (e.g. 'test-service.example.com' or 'test-service'). The environment variable takes precedence over the config file value.
```
Comment (ssteeg-mdsol) [inline README.md]: [nit] Why the change from `*` → `-`? The "Plugins" section is also still using `*`

File: README.md [hunk 4]
```diff
@@ -210,6 +217,29 @@ For example:
...
 ```
 
+## Utility classes
+
+### TraceWrapper
+
+This class provides a `.wrap_in_custom_span` method which expects a configuration hash, a span name and a block.
+You may also pass a span kind and an Application object using respectively `span_kind:` and `app:` keyword arguments.
+
+The block You pass will be executed in the context of a custom span.
```
Comment (ssteeg-mdsol) [inline README.md]: The block ~Y~you pass

File: README.md [hunk 5]
```diff
@@ -210,6 +217,29 @@ For example:
...
+This is useful when your application doesn't use the rack handler but still needs to generate complete traces, for instance background jobs or lambda calling remote services.
```
Comment (ssteeg-mdsol) [inline README.md]: lambda → lambdas

File: README.md (hunk 3 above)
Comment (jfeltesse-mdsol) [reply README.md]: 🙈 
just a habit. 

fixed in 698db93e6586dc8ebd33f2ff983a73e69e665785

File: lib/zipkin-tracer/trace_wrapper.rb (hunk 2 above)
Comment (jfeltesse-mdsol) [reply lib/zipkin-tracer/trace_wrapper.rb]: Great suggestions! Implemented in 8598529a56039cb08be9147b5a9aca3f0d230fcd

**Comment by jcarres-mdsol** [comment]
I guess similar code would be needed for other tracers using async, maybe can be isolated in a module or class. Why not make the change to all of them so we do not have the caviat of the that option only working on one?

Also, the sidekiq integration could be written now on top of the wrapper class?

**Comment by jfeltesse-mdsol** [comment]
> I guess similar code would be needed for other tracers using async, maybe can be isolated in a module or class. Why not make the change to all of them so we do not have the caviat of the that option only working on one?

time limit didn't allow me to go further.

> Also, the sidekiq integration could be written now on top of the wrapper class?

the topic of another PR, surely.

**Comment by jcarres-mdsol** [comment]
At least I think "async option to be applied to all traces which do async"  (maybe just one more tracer)  should be implemented in this PR, just to not leave it hanging. I see the point about sidekick

**Comment by jfeltesse-mdsol** [comment]
@jcarres-mdsol alright, done in 013332f. Enjoy!

File: lib/zipkin-tracer/zipkin_http_sender.rb [hunk 6]
```diff
@@ -32,13 +32,18 @@ class ZipkinHttpSender < ZipkinSenderBase
...
+      SuckerPunch.logger = options[:logger]
       super(options)
     end
 
     def flush!
-      AsyncHttpApiClient.perform_async(@json_api_host, spans.dup)
+      if @async
+        HttpApiClient.perform_async(@json_api_host, spans.dup)
+      else
+        HttpApiClient.new.perform(@json_api_host, spans.dup)
```
Comment (jcarres-mdsol) [inline lib/zipkin-tracer/zipkin_http_sender.rb]: I think we do not need to `.dup` here. Same comment for the other sender.
The reason we dup in async is to avoid the thread variable to be overwritten as the thread gets a new request while async is still to sent the spans of the previous request. In the sync case, that should not be an issue.

Comment (jfeltesse-mdsol) [reply lib/zipkin-tracer/zipkin_http_sender.rb]: done in f60e276
